tests/xoto3/dynamodb/write_versioned/transact/run_test.py

```diff
@@ -175,10 +175,82 @@ def test_delete(tx: VersionedTransaction) -> VersionedTransaction:
 
 
 def test_no_op_builder():
+    ran = False
+
     def builder(tx):
-        assert False
+        nonlocal ran
+        ran = True
+        # transaction does actually run even with no items specified,
+        # since lazy loading is now permitted.
         return tx
 
     versioned_transact_write_items(
         builder, dict(table1=[], table2=[]),
     )
+
+    assert ran
+
+
+def test_lazy_loading_reads_and_writes(
+    integration_test_id_table, integration_test_id_table_put, integration_test_id_table_cleaner
+):
+    tname = integration_test_id_table.name
+    test_id_source = "versioned-transact-known-item-read"
+    test_id_lazy = "versioned-transact-lazy-load-read"
+    test_dest_id = "versioned-transact-write-using-lazy-loaded-value"
+
+    integration_test_id_table_put(dict(id=test_id_source, val=10))
+    integration_test_id_table_put(dict(id=test_id_lazy, val=9))
+    integration_test_id_table_cleaner(dict(id=test_dest_id))
+
+    def lazy_op(tx: VersionedTransaction) -> VersionedTransaction:
+        src = require(tx, tname, dict(id=test_id_source))
+        if src["val"] > 5:
+            # the if statement here is just an example of why you might want to lazy-load something.
+            # In our test, this statement always passes because of the fixture data.
+            lazy = require(tx, tname, dict(id=test_id_lazy))
+            print(lazy)
+            dest_item = dict(id=test_dest_id, val=src["val"] + lazy["val"])
+            print(dest_item)
+            return put(tx, tname, dest_item)
+        # this part of the test is just an example of what you might otherwise do.
+        # it's not actually ever going to run in our test.
+        return tx
+
+    # note that we only specify upfront a key for the single item we know we need to prefetch
+    result = versioned_transact_write_items(lazy_op, {tname: [dict(id=test_id_source)]},)
+
+    assert require(result, tname, dict(id=test_dest_id)) == dict(id=test_dest_id, val=19)
+
+
+def test_optimistic_delete_nonexistent(integration_test_id_table):
+    test_id_to_delete = "versioned-transact-opt-delete"
+
+    def opt_delete(tx: VersionedTransaction) -> VersionedTransaction:
+        return delete(tx, integration_test_id_table.name, dict(id=test_id_to_delete))
+
+    res = versioned_transact_write_items(opt_delete, dict())
+
+    assert None is get(res, integration_test_id_table.name, dict(id=test_id_to_delete))
```

Comment:
Why `None is` instead of `is None`?

Reply:
i tend to think that putting the "what" on the left side and the "how" on the right makes things more readable.

this follows the general pattern of most languages and their code.

i'm just not all that consistent about it.

Reply:
Interesting. The general pattern I thing I've seen is more `actual` on the left and `expected` on the right.

Reply:
i've seen that too, i'm just not sure it makes as much sense. it doesn't fit with the way we use programming languages in general.

there's a big exception to this rule that you may have encountered before:

```
expected = dict(foo='bar')
assert expected = do_the_actual(...)
```

I would not advocate putting variable names on the left, because you're risking accidentally assigning and thinking that your test passes when it doesn't.

but if you're writing small tests (which i often don't do), there's likely no need for the intermediate variable, and you could just as safely write

```
assert dict(foo='bar') == do_the_actual(...)
```

Reply:
in the end, as long as it's not ambiguous what you're testing, i'm not going to complain if you do it the other way.